server/lock.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"encoding/json"
+	"regexp"
+	"time"
+
+	"github.com/mattermost/mattermost-server/model"
+)
+
+const (
+	// LOCK_KEY is used to prevent multiple instances of the plugin from scheduling surveys in parallel.
+	LOCK_KEY = "Lock"
+
+	// USER_LOCK_KEY is used to prevent multiple instances of the plugin from responding to a single user's requests
+	// in parallel.
+	USER_LOCK_KEY = "UserLock-%s"
+
+	// LOCK_EXPIRATION is how long a lock can be held before it will be automatically released the next time that an
+	// instance of the plugin is started up.
+	LOCK_EXPIRATION = time.Hour
+)
+
+var userLockPattern = regexp.MustCompile("^UserLock-.{26}$")
+
+func (p *Plugin) tryLock(key string, now time.Time) (bool, *model.AppError) {
+	b, err := json.Marshal(now)
+	if err != nil {
+		return false, &model.AppError{Message: err.Error()}
+	}
+
+	return p.API.KVCompareAndSet(key, nil, b)
+}
+
+func (p *Plugin) unlock(key string) *model.AppError {
+	return p.API.KVDelete(key)
+}
+
+// clearStaleLocks deletes any lock entries that have been held for a long time since that likely means that the routine
+// that held them died without properly releasing them.
+func (p *Plugin) clearStaleLocks(now time.Time) *model.AppError {
```

Comment:
I'm wondering why this happens. Why should a go routine die? Shouldn't a `deffer` take care of this?

Reply:
The defer statements should release the locks properly, but I wanted to be extra careful in case the process gets killed by something external because that could cause a deadlock in the plugin code which would be very difficult to debug.

Reply:
I see. Thanks for the explanation.